### PR TITLE
pubsub: remove looping logic from getNextBatch to eliminate a race condition

### LINKED ISSUE #1238: pubsub/tests: TestReceiveReturnsErrorFromSendAcks is flaky
Test failed on Travis:

```
--- FAIL: TestReceiveReturnsErrorFromSendAcks (5.02s)
    acks_test.go:341: got context deadline exceeded; last error: <nil>, want SendAcks failed
FAIL
```

I couldn't repro locally.

### PR DESCRIPTION
**Pull request by ijt**
Fixes #1238.

There was a race condition: If `m.Ack()` results in a permanent error, that should be reported from the next call to `sub.Receive()` but if `sub.Receive()` acquires the subscription's mutex before that permanent error occurs then the call to `getNextBatch()` goes into an infinite loop waiting for a non-empty batch. This change replaces the empty batch loop in `getNextBatch()` with just its body so that it will return and the check for unrecoverable errors within `Receive()` can be reached again even if no more messages come in.

Before:
```
[ ~/src/gocloud.dev/pubsub ] go test -run=TestReceiveReturnsErrorFromSendAcks -count=10000
--- FAIL: TestReceiveReturnsErrorFromSendAcks (5.00s)
    acks_test.go:341: got context deadline exceeded; last error: <nil>, want SendAcks failed
--- FAIL: TestReceiveReturnsErrorFromSendAcks (5.00s)
    acks_test.go:341: got context deadline exceeded; last error: <nil>, want SendAcks failed
FAIL
exit status 1
FAIL	gocloud.dev/pubsub	10.658s
```
After:
```
[ ~/src/gocloud.dev/pubsub ] go test -run=TestReceiveReturnsErrorFromSendAcks -count=100000
PASS
ok  	gocloud.dev/pubsub	5.469s
```

### CODE REVIEW

File: pubsub/pubsub.go [hunk 1]
```diff
@@ -423,7 +423,15 @@ func (s *Subscription) Receive(ctx context.Context) (_ *Message, err error) {
 func (s *Subscription) getNextBatch(ctx context.Context, nMessages int) ([]*Message, error) {
 	var msgs []*driver.Message
 	for len(msgs) == 0 {
-		err := retry.Call(ctx, gax.Backoff{}, s.driver.IsRetryable, func() error {
+		// Allow unrecoverable errors from acks to be detected and reported.
```
Comment (vangent) [inline pubsub/pubsub.go]: What would happen if we allowed `getNextBatch` to return an empty slice instead? Wouldn't we just loop in `Receive`?

E.g., try commenting out your new code, and also deleting the `for len(msgs) == 0` loop in this function, does it wor?

File: pubsub/pubsub.go [hunk 2]
```diff
@@ -423,7 +423,15 @@ func (s *Subscription) Receive(ctx context.Context) (_ *Message, err error) {
...
+		s.mu.Lock()
+		err := s.err
+		s.mu.Unlock()
+		if s.err != nil {
```
Comment (jba) [inline pubsub/pubsub.go]: use `err` instead of `s.err` here and the line below.

Comment (ijt) [reply pubsub/pubsub.go]: Done.

File: pubsub/pubsub.go (hunk 1 above)
Comment (ijt) [reply pubsub/pubsub.go]: It does appear to work. Good idea.

File: pubsub/pubsub.go (hunk 2 above)
Comment (ijt) [reply pubsub/pubsub.go]: Obsoleted by @vangent's idea above.

**Comment by ijt** [comment]
PTAL

**Comment by ijt** [comment]
@vangent, done.

**Pull request merged**